### DISCUSSION TITLE
Cross-namespace endpoints

### DISCUSSION THREAD

**Discussion by nrfox** [kiali/kiali#4801] [Roadmap]
Some API endpoints in kiali are scoped by namespace, such as [services](https://github.com/kiali/kiali/blob/master/routing/routes.go#L328) and [workloads](https://github.com/kiali/kiali/blob/master/routing/routes.go#L556). A consequence of this is that the UI must make a request _per namespace_ to gather information from all the observed and selected namespaces. As the number of namespaces grows, the number of requests grows from the UI.

Other API endpoints in kiali are already cross-namespace such as the [tls](https://github.com/kiali/kiali/blob/master/routing/routes.go#L1003) endpoint that fetches the tls status for the whole mesh.

I propose that kiali adds a cross-namespace endpoint for each of the endpoints where the UI needs to gather info across namespaces. Specifically:
- workloads
- services
- apps

Pros:
- Reduces n requests to the kiali API (where n == number of namespaces) down to 1 request. This is less about reducing load on the kiali api and more about reducing the round trip time required to make each request. Browsers can only make a limited amount of requests in parallel and each additional request you make also increases the chances that you'll have an outlier that is slower than the rest. Loading of pages in Kiali is blocked until the slowest request returns.
- Greater flexibility for filtering requests. The UI could query across namespaces with a regex or label filter and the backend could handle the details of making this work with the kube API. This opens the door for more options to filter objects on the front end.

Cons:
- A single cross-namespace request may be slower than a single namespaced request for info even if the cross-namespace request is much faster than the aggregate namespaced requests.

Notes:
[This](https://github.com/kiali/kiali/pull/4796) POC to provide a cross-namespace endpoint to fetch the istio config is related to this discussion.

**Comment by nrfox** [comment]
cc @israel-hdez @aljesusg as a follow-up from previous health refactoring discussions

**Comment by lucasponce** [comment]
Thanks, @nrfox, as you pointed @hhovsepy is working in the same area with the Istio endpoints.

In general, yes, with the perspective of the lessons learned, I do think that cross-namespace validations would at least simplify the design without adding overhead, as some of the promises wait until all the namespaces are resolved.

Also, on the other hand, the k8s API is not designed to be cross-namespace and that may introduce some limitations in future integrations; I don't think they will affect the short-term of Kiali, but potentially it may add some constraints (I'm just introducing this comment here, as probably some of my feedback may be biased for that aspect).

But I'm +1 to start thinking on this for future iterations.

I'd start in small steps to have under control all the refactoring.

**Comment by hhovsepy** [comment]
Thanks @nrfox and @lucasponce , we already had an resolved issue https://github.com/kiali/kiali/issues/4692 about having endpoint to retrieve all configs from registry, and using it in Service details.
For further improvements feel we can have more issues logged covering different areas, feel free the add them.

**Comment by nrfox** [comment]
> Also, on the other hand, the k8s API is not designed to be cross-namespace and that may introduce some limitations in future integrations; I don't think they will affect the short-term of Kiali, but potentially it may add some constraints (I'm just introducing this comment here, as probably some of my feedback may be biased for that aspect).

Although it is limited in what you can do cross-namespace, the k8s API does have some support for listing objects across namespaces. GVRs have a high level endpoint you can query to list all objects in all namespaces e.g. `/api/v1/services` to list all services. You can see this if you crank up the verbosity with kubectl:
```
kubectl get svc -v 6 -A
```

Where we have run into issues with the kube api afaik is wanting to query all objects across a limited number of namespaces. Something like `/api/v1/services?namespaces=default,bookinfo,ns1,ns2`. The kube informers specifically do not support that. However, the kiali API could support that and we can figure out what the communication looks like with the kube api on the backend.

Example querying workloads for specific namespaces:
`/api/workloads?namespaces=default,bookinfo,ns1,ns2`
Regex:
`/api/workloads?namespaces=*.bookinfo`

There's only so many query params you can fit into a request url so we'll probably need to handle that edge case on the front end since I'm not sure there is a limit on the number of namespaces you can have.

I'm happy to write up a KEP to go into more detail.

**Reply by lucasponce** [reply]
Yes, but the --all-namespaces semantic requires cluster rights :) that was the main limitation of all the designs we have in place in the backend.

Don't get me wrong, I'm +1 with passing the namespaces as parameters, I guess the change should be minimal to the "handlers" package that can iterate over the "business" one, and that can be the first step.

I guess the change is clear, no need to create a KEP if you don't want, with just an issue/epic could be enough.
